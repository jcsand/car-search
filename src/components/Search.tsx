/** @jsxImportSource @emotion/react */
import React from "react";
import styled from "@emotion/styled";

import { Button } from "@@components/Button";
import { SearchLocation } from "@@components/SearchLocation";
import { mediaQuery, t } from "@@lib/themeHelpers";

const mq = mediaQuery(["l"]);

const SearchContainer = styled.div`
  display: flex;
  position: relative;
  margin: ${t.space(4)} 0;
  padding: ${t.space(1)};
  border-radius: ${t.space(3)};
  background: ${t.color("lightOrange")};
  box-shadow: 0 ${t.space(1, true)} ${t.space(3)} ${t.color("black", 16 / 100)};

  ${mq({
    flexDirection: ["column", "row"]
  })}
`;

const SearchButton = styled(Button)`
  height: ${t.space(12)};
  padding: ${t.space(2)} ${t.space(4)};
  background-color: ${t.color("green")};
  font-size: ${t.fontSize(3)};
  font-weight: 700;
  line-height: ${t.lineHeight(4)};

  &:focus,
  &:hover {
    background-color: ${t.color("darkGreen")};
  }

  &:focus {
    outline-style: none;
    box-shadow: 0 0 0 3px ${t.color("lightBlue", 24 / 100)};
  }

  ${mq((t) => ({
    borderRadius: [
      `${t.space(2, true)} ${t.space(2, true)} ${t.space(2)} ${t.space(2)}`,
      `${t.space(2, true)} ${t.space(2)} ${t.space(2)} ${t.space(2, true)}`
    ]
  }))}
`;

export const Search: React.FC = () => {
  return (
    <SearchContainer>
      <SearchLocation />
      <SearchButton>Search</SearchButton>
    </SearchContainer>
  );
};
