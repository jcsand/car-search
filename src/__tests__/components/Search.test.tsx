import React from "react";
import { fireEvent, waitFor, screen } from "@testing-library/react";
import { getAllByRole } from "@testing-library/dom";
import "@testing-library/jest-dom";
import { mockRequests, renderWithTheme } from "../__helpers__";
import { Search } from "@@components/Search";

describe("Search component", () => {
  const expectedPlaceholder = "Pick-up Location";

  mockRequests();

  it("renders matching snapshot", async () => {
    const rendered = renderWithTheme(<Search />);
    expect(rendered.baseElement).toMatchSnapshot();
  });

  it("renders with results matching snapshot", async () => {
    const rendered = renderWithTheme(<Search />);

    await waitFor(() => screen.getByLabelText(expectedPlaceholder));
    const searchInput = screen.getByLabelText(expectedPlaceholder);

    await fireEvent.change(searchInput, { target: { value: "manchester" } });
    await waitFor(() => screen.getByRole("listbox"));

    expect(rendered.baseElement).toMatchSnapshot();
  });

  it("replaces search query on suggestion click", async () => {
    renderWithTheme(<Search />);

    await waitFor(() => screen.getByLabelText(expectedPlaceholder));
    const searchInput = screen.getByLabelText(expectedPlaceholder);

    await fireEvent.change(searchInput, { target: { value: "manchester" } });
    await waitFor(() => screen.getByRole("listbox"));

    const firstSuggestion = screen.getByRole("listbox").children[0];
    await fireEvent.click(firstSuggestion, new MouseEvent("click"));

    const expectedValue =
      "Manchester Airport (MAN), Manchester, United Kingdom";
    await waitFor(() => screen.getByDisplayValue(expectedValue));

    expect(searchInput).toHaveValue(expectedValue);
  });

  it("navigates the search suggestions when using arrow keys", async () => {
    renderWithTheme(<Search />);

    const searchInput = await waitFor(() =>
      screen.getByLabelText(expectedPlaceholder)
    );

    await fireEvent.change(searchInput, { target: { value: "manchester" } });
    const suggestions = await waitFor(() => screen.getByRole("listbox"));
    const options = getAllByRole(suggestions, "option");

    await fireEvent.keyDown(searchInput, { key: "ArrowDown" });
    await waitFor(() => {
      expect(options[0]).toHaveAttribute("aria-selected", "true");
    });

    await fireEvent.keyDown(searchInput, { key: "ArrowDown" });
    await waitFor(() =>
      expect(options[1]).toHaveAttribute("aria-selected", "true")
    );
    expect(options[0]).toHaveAttribute("aria-selected", "false");

    await fireEvent.keyDown(searchInput, { key: "ArrowUp" });
    await waitFor(() =>
      expect(options[0]).toHaveAttribute("aria-selected", "true")
    );
    expect(options[1]).toHaveAttribute("aria-selected", "false");

    await fireEvent.keyDown(searchInput, { key: "ArrowUp" });
    await waitFor(() =>
      expect(options[0]).toHaveAttribute("aria-selected", "false")
    );
    expect(options[1]).toHaveAttribute("aria-selected", "false");
  });

  it("should select the currently selected suggestion on enter press", async () => {
    renderWithTheme(<Search />);

    const searchInput = await waitFor(() =>
      screen.getByLabelText(expectedPlaceholder)
    );

    await fireEvent.change(searchInput, { target: { value: "manchester" } });
    const suggestions = await waitFor(() => screen.getByRole("listbox"));
    const options = getAllByRole(suggestions, "option");

    await fireEvent.keyDown(searchInput, { key: "ArrowDown" });
    await waitFor(() => {
      expect(options[0]).toHaveAttribute("aria-selected", "true");
    });

    await fireEvent.keyDown(searchInput, { key: "Enter" });
    await waitFor(() => expect(() => screen.getByRole("listbox")).toThrow());

    const expectedValue =
      "Manchester Airport (MAN), Manchester, United Kingdom";
    expect(searchInput).toHaveValue(expectedValue);
  });
});
