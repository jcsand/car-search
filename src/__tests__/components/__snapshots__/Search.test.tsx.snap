// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders matching snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  margin: 16px 0;
  padding: 4px;
  border-radius: 12px;
  background: #ffb700;
  box-shadow: 0 2px 12px rgba(0, 0, 0, 0.16);
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width: 1024px) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0 0 4px 0;
}

.emotion-1::before {
  content: "";
  display: inline-block;
  position: absolute;
  z-index: 1;
  top: 12px;
  left: 16px;
  width: 16px;
  height: 16px;
  margin-top: 12px;
  margin-right: 8px;
  background: url("test-file-stub");
  -webkit-background-position: center;
  background-position: center;
  -webkit-background-size: cover;
  background-size: cover;
  pointer-events: none;
}

@media (min-width: 1024px) {
  .emotion-1 {
    margin: 0 4px 0 0;
  }
}

.emotion-2 {
  width: 100%;
  height: 64px;
  padding: 8px;
  padding-left: 44px;
  border: none;
  color: #1a1a1a;
  font-family: BlinkMacSystemFont,-apple-system,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  border-radius: 8px 8px 2px 2px;
}

.emotion-2:focus {
  outline-style: none;
  box-shadow: 0 0 0 3px rgba(18, 115, 196, 0.24);
}

.emotion-2::-webkit-input-placeholder {
  color: #949494;
}

.emotion-2::-moz-placeholder {
  color: #949494;
}

.emotion-2:-ms-input-placeholder {
  color: #949494;
}

.emotion-2::placeholder {
  color: #949494;
}

@media (min-width: 1024px) {
  .emotion-2 {
    border-radius: 8px 2px 2px 8px;
  }
}

.emotion-3 {
  padding: 8px;
  border: 1px solid transparent;
  color: #fff;
  font-family: "Avenir Next",BlinkMacSystemFont,-apple-system,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-size: 14px;
  line-height: 12px;
  cursor: pointer;
  height: 64px;
  padding: 8px 16px;
  background-color: #068323;
  font-size: 20px;
  font-weight: 700;
  line-height: 28px;
  border-radius: 4px 4px 8px 8px;
}

.emotion-3:focus {
  outline: none;
  box-shadow: 0 0 0 3px rgba(255, 255, 255, 0.24);
}

.emotion-3:focus,
.emotion-3:hover {
  background-color: #0d6521;
}

.emotion-3:focus {
  outline-style: none;
  box-shadow: 0 0 0 3px rgba(18, 115, 196, 0.24);
}

@media (min-width: 1024px) {
  .emotion-3 {
    border-radius: 4px 8px 8px 4px;
  }
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <input
          aria-label="Pick-up Location"
          class="emotion-2"
          placeholder="Pick-up Location"
          value=""
        />
      </div>
      <button
        class="emotion-3"
      >
        Search
      </button>
    </div>
  </div>
</body>
`;

exports[`Search component renders with results matching snapshot 1`] = `
@keyframes animation-0 {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  margin: 16px 0;
  padding: 4px;
  border-radius: 12px;
  background: #ffb700;
  box-shadow: 0 2px 12px rgba(0, 0, 0, 0.16);
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width: 1024px) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0 0 4px 0;
}

.emotion-1::before {
  content: "";
  display: inline-block;
  position: absolute;
  z-index: 1;
  top: 12px;
  left: 16px;
  width: 16px;
  height: 16px;
  margin-top: 12px;
  margin-right: 8px;
  background: url("test-file-stub");
  -webkit-background-position: center;
  background-position: center;
  -webkit-background-size: cover;
  background-size: cover;
  pointer-events: none;
}

@media (min-width: 1024px) {
  .emotion-1 {
    margin: 0 4px 0 0;
  }
}

.emotion-2 {
  width: 100%;
  height: 64px;
  padding: 8px;
  padding-left: 44px;
  border: none;
  color: #1a1a1a;
  font-family: BlinkMacSystemFont,-apple-system,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  border-radius: 8px 8px 2px 2px;
}

.emotion-2:focus {
  outline-style: none;
  box-shadow: 0 0 0 3px rgba(18, 115, 196, 0.24);
}

.emotion-2::-webkit-input-placeholder {
  color: #949494;
}

.emotion-2::-moz-placeholder {
  color: #949494;
}

.emotion-2:-ms-input-placeholder {
  color: #949494;
}

.emotion-2::placeholder {
  color: #949494;
}

@media (min-width: 1024px) {
  .emotion-2 {
    border-radius: 8px 2px 2px 8px;
  }
}

.emotion-3 {
  position: absolute;
  top: calc(100% + 4px);
  width: 100%;
  border: none;
  border-radius: 4px;
  border-color: #1a1a1a;
  background: #fff;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.16);
  color: #1a1a1a;
  -webkit-animation: animation-0 0.5s;
  animation: animation-0 0.5s;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px;
  cursor: pointer;
}

.emotion-4:hover,
.emotion-4[aria-selected="true"] {
  background: rgba(18, 115, 196, 0.06);
}

.emotion-5 {
  display: inline-block;
  min-width: 5.1rem;
  padding: 0.3rem 0;
  border: none;
  border-radius: 4px;
  color: #fff;
  font-size: 14px;
  font-weight: 500;
  line-height: 1rem;
  text-align: center;
  background: #ff8000;
  color: #262626;
}

.emotion-6 {
  padding-left: 12px;
}

.emotion-7 {
  font-family: BlinkMacSystemFont,-apple-system,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
}

.emotion-8 {
  color: #474747;
  font-family: BlinkMacSystemFont,-apple-system,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-size: 12px;
  font-weight: 500;
  line-height: 18px;
}

.emotion-10 {
  display: inline-block;
  min-width: 5.1rem;
  padding: 0.3rem 0;
  border: none;
  border-radius: 4px;
  color: #fff;
  font-size: 14px;
  font-weight: 500;
  line-height: 1rem;
  text-align: center;
  background: #0071c2;
}

.emotion-34 {
  padding: 8px;
  border: 1px solid transparent;
  color: #fff;
  font-family: "Avenir Next",BlinkMacSystemFont,-apple-system,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-size: 14px;
  line-height: 12px;
  cursor: pointer;
  height: 64px;
  padding: 8px 16px;
  background-color: #068323;
  font-size: 20px;
  font-weight: 700;
  line-height: 28px;
  border-radius: 4px 4px 8px 8px;
}

.emotion-34:focus {
  outline: none;
  box-shadow: 0 0 0 3px rgba(255, 255, 255, 0.24);
}

.emotion-34:focus,
.emotion-34:hover {
  background-color: #0d6521;
}

.emotion-34:focus {
  outline-style: none;
  box-shadow: 0 0 0 3px rgba(18, 115, 196, 0.24);
}

@media (min-width: 1024px) {
  .emotion-34 {
    border-radius: 4px 8px 8px 4px;
  }
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <input
          aria-label="Pick-up Location"
          class="emotion-2"
          placeholder="Pick-up Location"
          value="manchester"
        />
        <ul
          class="emotion-3"
          role="listbox"
        >
          <li
            aria-selected="false"
            class="emotion-4"
            id="search-suggestion-item-0"
            role="option"
          >
            <span
              class="emotion-5"
            >
              Airport
            </span>
            <div
              class="emotion-6"
            >
              <p
                class="emotion-7"
              >
                Manchester Airport (MAN)
              </p>
              <p
                class="emotion-8"
              >
                Manchester, Greater Manchester, United Kingdom
              </p>
            </div>
          </li>
          <li
            aria-selected="false"
            class="emotion-4"
            id="search-suggestion-item-1"
            role="option"
          >
            <span
              class="emotion-10"
            >
              City
            </span>
            <div
              class="emotion-6"
            >
              <p
                class="emotion-7"
              >
                Manchester
              </p>
              <p
                class="emotion-8"
              >
                Greater Manchester, United Kingdom
              </p>
            </div>
          </li>
          <li
            aria-selected="false"
            class="emotion-4"
            id="search-suggestion-item-2"
            role="option"
          >
            <span
              class="emotion-10"
            >
              Station
            </span>
            <div
              class="emotion-6"
            >
              <p
                class="emotion-7"
              >
                Manchester - Piccadilly Train Station
              </p>
              <p
                class="emotion-8"
              >
                Manchester, England, United Kingdom
              </p>
            </div>
          </li>
          <li
            aria-selected="false"
            class="emotion-4"
            id="search-suggestion-item-3"
            role="option"
          >
            <span
              class="emotion-10"
            >
              City
            </span>
            <div
              class="emotion-6"
            >
              <p
                class="emotion-7"
              >
                Oldham
              </p>
              <p
                class="emotion-8"
              >
                Greater Manchester, United Kingdom
              </p>
            </div>
          </li>
          <li
            aria-selected="false"
            class="emotion-4"
            id="search-suggestion-item-4"
            role="option"
          >
            <span
              class="emotion-10"
            >
              City
            </span>
            <div
              class="emotion-6"
            >
              <p
                class="emotion-7"
              >
                Bolton
              </p>
              <p
                class="emotion-8"
              >
                Greater Manchester, United Kingdom
              </p>
            </div>
          </li>
          <li
            aria-selected="false"
            class="emotion-4"
            id="search-suggestion-item-5"
            role="option"
          >
            <span
              class="emotion-10"
            >
              City
            </span>
            <div
              class="emotion-6"
            >
              <p
                class="emotion-7"
              >
                Stockport
              </p>
              <p
                class="emotion-8"
              >
                Greater Manchester, United Kingdom
              </p>
            </div>
          </li>
        </ul>
      </div>
      <button
        class="emotion-34"
      >
        Search
      </button>
    </div>
  </div>
</body>
`;
